libcxx/include/__ranges/concat_view.h

```diff
@@ -0,0 +1,637 @@
+        satisfy<_Idx + 1>();
+      }
+    }
+  }
+
+  template <std::size_t _Idx>
+  _LIBCPP_HIDE_FROM_ABI constexpr void prev() {
+    if constexpr (_Idx == 0) {
+      --std::get<0>(it_);
+    } else {
+      if (std::get<_Idx>(it_) == ranges::begin(std::get<_Idx>(parent_->__views_))) {
+        using prev_view = __maybe_const<_Const, tuple_element_t<_Idx - 1, tuple<_Views...>>>;
+        if constexpr (common_range<prev_view>) {
+          it_.template emplace<_Idx - 1>(ranges::end(std::get<_Idx - 1>(parent_->__views_)));
+        } else {
+          it_.template emplace<_Idx - 1>(ranges::__next(ranges::begin(std::get<_Idx - 1>(parent_->__views_)),
+                                                        ranges::size(std::get<_Idx - 1>(parent_->__views_))));
+        }
+        prev<_Idx - 1>();
+      } else {
+        --std::get<_Idx>(it_);
+      }
+    }
+  }
+
+  template <std::size_t _Idx>
+  _LIBCPP_HIDE_FROM_ABI constexpr void advance_fwd(difference_type __offset, difference_type __steps) {
+    using underlying_diff_type = iter_difference_t<variant_alternative_t<_Idx, base_iter>>;
+    if constexpr (_Idx == sizeof...(_Views) - 1) {
+      std::get<_Idx>(it_) += static_cast<underlying_diff_type>(__steps);
+    } else {
+      difference_type __n_size = ranges::size(std::get<_Idx>(parent_->__views_));
+      if (__offset + __steps < __n_size) {
+        std::get<_Idx>(it_) += static_cast<underlying_diff_type>(__steps);
+      } else {
+        it_.template emplace<_Idx + 1>(ranges::begin(std::get<_Idx + 1>(parent_->__views_)));
+        advance_fwd<_Idx + 1>(0, __offset + __steps - __n_size);
+      }
+    }
+  }
+
+  template <std::size_t _Idx>
+  _LIBCPP_HIDE_FROM_ABI constexpr void advance_bwd(difference_type __offset, difference_type __steps) {
+    using underlying_diff_type = iter_difference_t<variant_alternative_t<_Idx, base_iter>>;
+    if constexpr (_Idx == 0) {
+      std::get<_Idx>(it_) -= static_cast<underlying_diff_type>(__steps);
+    } else {
+      if (__offset >= __steps) {
+        std::get<_Idx>(it_) -= static_cast<underlying_diff_type>(__steps);
+      } else {
+        auto __prev_size = ranges::__distance(std::get<_Idx - 1>(parent_->__views_));
+        it_.template emplace<_Idx - 1>(ranges::begin(std::get<_Idx - 1>(parent_->__views_)) + __prev_size);
+        advance_bwd<_Idx - 1>(__prev_size, __steps - __offset);
+      }
+    }
+  }
+
+  template <size_t... _Is, typename _Func>
+  _LIBCPP_HIDE_FROM_ABI constexpr void
+  apply_fn_with_const_index(size_t __index, _Func&& __func, std::index_sequence<_Is...>) {
+    ((__index == _Is ? (__func(std::integral_constant<size_t, _Is>{}), 0) : 0), ...);
+  }
+
+  template <size_t _Idx, typename _Func>
+  _LIBCPP_HIDE_FROM_ABI constexpr void apply_fn_with_const_index(size_t __index, _Func&& __func) {
+    apply_fn_with_const_index(__index, std::forward<_Func>(__func), std::make_index_sequence<_Idx>{});
+  }
```

Comment:
`apply_fn_with_const_index` is not __uglified. `index_sequence`, `integral_constant`, and `make_index_sequence` shouldn't be `std::`-qualified here (as they are not functions).

Also, do `const` and `fn` in this name make sense? I think we can just use `__apply_at_index`.